File: Algorithm/src/main/java/Acwing/Ac827_双链表/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scanner = bufio.NewScanner(os.Stdin)

func nextInt() int {
	scanner.Scan()
	ans, _ := strconv.Atoi(scanner.Text())
	return ans
}

func nextStr() string {
	scanner.Scan()
	return scanner.Text()
}

type ListNode struct {
	val       int
	pre, next *ListNode
}

// --------------
var kthMap = make(map[int]*ListNode)
var fakeHead = ListNode{val: -1}
var fakeTail = ListNode{val: -1}
var time = 0

func incr(node *ListNode) {
	time++
	kthMap[time] = node
}
func addNode(pre, cur, next *ListNode) {
	pre.next = cur
	cur.pre = pre
	cur.next = next
	next.pre = cur
}

func delNode(pre, cur, next *ListNode) {
	pre.next = next
	next.pre = pre
	cur.pre = nil
	cur.next = nil
}

func printList() {
	cur := fakeHead.next
	for cur != &fakeTail {
		fmt.Printf("%d ", cur.val)
		cur = cur.next
	}
	fmt.Println()
}

func init() {
	scanner.Split(bufio.ScanWords)
	fakeTail.pre = &fakeHead
	fakeHead.next = &fakeTail
}
func main() {
	n := nextInt()
	for i := 0; i < n; i++ {
		target := nextStr()
		switch target {
		case "L":
			x := nextInt()
			futureHead := ListNode{val: x}
			curHead := fakeHead.next
			addNode(&fakeHead, &futureHead, curHead)
			incr(&futureHead)
			break
		case "R":
			x := nextInt()
			futureTail := ListNode{val: x}
			curTail := fakeTail.pre
			addNode(curTail, &futureTail, &fakeTail)
			incr(&futureTail)
			break
		case "D":
			k := nextInt()
			toDelNode := kthMap[k]
			delNode(toDelNode.pre, toDelNode, toDelNode.next)
			break
		case "IL":
			k := nextInt()
			x := nextInt()
			kthNode := kthMap[k]
			kthPreNode := kthNode.pre
			toAddNode := ListNode{val: x}
			addNode(kthPreNode, &toAddNode, kthNode)
			incr(&toAddNode)
			break
		case "IR":
			k := nextInt()
			x := nextInt()
			kthNode := kthMap[k]
			kthNextNode := kthNode.next
			toAddNode := ListNode{val: x}
			addNode(kthNode, &toAddNode, kthNextNode)
			incr(&toAddNode)
			break
		default:
			break
		}
	}
	printList()
}

File: Algorithm/src/main/java/Acwing/Ac827_双链表/Main.java
package Acwing.Ac827_双链表;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.HashMap;

class ListNode{
    int val;
    ListNode pre;
    ListNode next;

    public ListNode(int val) {
        this.val = val;
    }
}

public class Main {
    static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    static StreamTokenizer st = new StreamTokenizer(in);
    static int nextInt() throws IOException {
        st.nextToken();
        return (int) st.nval;
    }

    static String nextStr() throws IOException {
        st.nextToken();
        return st.sval;
    }

    static HashMap<Integer,ListNode> map = new HashMap<>();
    static ListNode fakeHead = new ListNode(Integer.MIN_VALUE), fakeTail = new ListNode(Integer.MAX_VALUE);
    static int time = 0;
    static {
        fakeHead.next = fakeTail;
        fakeTail.pre = fakeHead;
    }

    static void incr(ListNode node){
        time++;
        map.put(time,node);
    }

    static void printList(){
        ListNode cur = fakeHead;
        while ((cur = cur.next)!=fakeTail){
            System.out.print(cur.val+" ");
        }
        System.out.println();
    }

    static void addNode(ListNode pre,ListNode cur,ListNode next){
        pre.next = cur;
        cur.next = next;
        next.pre = cur;
        cur.pre = pre;
    }

    static void delNode(ListNode pre,ListNode cur,ListNode next){
        pre.next = next;
        next.pre = pre;
        cur.next = null;
        cur.pre = null;
    }

    public static void main(String[] args) throws IOException {
        int n = nextInt();
        for(int i=0;i<n;i++){
            String target = nextStr();
            int x,k;
            switch (target){
                case "L":
                    x = nextInt();
                    ListNode curHead = fakeHead.next;
                    ListNode futureHead = new ListNode(x);
                    addNode(fakeHead,futureHead,curHead);
                    incr(futureHead);
                    break;
                case "R":
                    x = nextInt();
                    ListNode curTail = fakeTail.pre;
                    ListNode futureTail = new ListNode(x);
                    addNode(curTail,futureTail,fakeTail);
                    incr(futureTail);
                    break;
                case "D":
                    k = nextInt();
                    ListNode toDelNode = map.get(k);
                    ListNode preNode = toDelNode.pre;
                    ListNode nextNode = toDelNode.next;
                    delNode(preNode,toDelNode,nextNode);
                    break;
                case "IL":
                    k = nextInt();x=nextInt();
                    ListNode kthNode = map.get(k);
                    ListNode kthPreNode = kthNode.pre;
                    ListNode toAddNode = new ListNode(x);
                    addNode(kthPreNode,toAddNode,kthNode);
                    incr(toAddNode);
                    break;
                case "IR":
                    k = nextInt();x=nextInt();
                    ListNode kthNode2 = map.get(k);
                    ListNode kthAftNode2 = kthNode2.next;
                    ListNode toAddNode2 = new ListNode(x);
                    addNode(kthNode2,toAddNode2,kthAftNode2);
                    incr(toAddNode2);
                    break;
                default:
                    break;
            }
        }
        printList();

    }


}
